libbeat/scripts/cmd/global_fields/main.go

```diff
@@ -39,40 +39,33 @@ func main() {
 		os.Exit(1)
 	}
 
-	err := os.MkdirAll(filepath.Join(*beatPath, "_meta"), 0755)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Cannot create _meta dir for %s: %+v\n", name, err)
-		os.Exit(1)
-	}
-
-	if len(beatFieldsPaths) == 0 {
+	if len(beatFieldsPaths) == 0 && *esBeatsPath == *beatPath {
 		fmt.Println("No field files to collect")
-		err = fields.AppendFromLibbeat(*esBeatsPath, *beatPath)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Cannot generate global fields.yml for %s: %+v\n", name, err)
-			os.Exit(2)
-		}
 		return
 	}
 
 	var fieldsFiles []*fields.YmlFile
 	for _, fieldsFilePath := range beatFieldsPaths {
-		pathToModules := filepath.Join(*beatPath, fieldsFilePath)
-
-		fieldsFile, err := fields.CollectModuleFiles(pathToModules)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Cannot collect fields.yml files: %+v\n", err)
-			os.Exit(2)
-		}
-
-		fieldsFiles = append(fieldsFiles, fieldsFile...)
+		fieldsFiles = collectFilesFromDir(*beatPath, fieldsFilePath, fieldsFiles)
 	}
 
-	err = fields.Generate(*esBeatsPath, *beatPath, fieldsFiles)
+	err := fields.Generate(*esBeatsPath, *beatPath, fieldsFiles)
```

Comment:
How about internalizing the collection of module files to the `fields.Generate()`. Make it accept a list of module directories like `Generate(elasticBeatsDir, beatDir string, moduleDirs ...string) ([]byte, error)`.

Note that I added a `[]byte` return which is the actual content. Then the caller becomes responsible for writing the content which makes it a bit more reusable. I think the YAML content is small enough to be rendered to memory without any issues.